File: cpp/_410.cpp
// O(NlogM) / O(1)
class Solution {
public:
    int splitArray(vector<int>& nums, int m) {
        int l = *max_element(nums.begin(), nums.end());
        int r = accumulate(nums.begin(), nums.end(), 0);
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (count(nums, mid) <= m)
                r = mid;
            else
                l = mid + 1;
        }
        return l;
    }
private:
    int count(vector<int>& nums, int d) {
        int ans = 1, cur = 0;
        for (auto& num : nums) {
            cur += num;
            if (cur > d) {
                ans ++;
                cur = num;
            }
        }
        return ans;
    }
};

File: python/_410.py
# O(NlogM) / O(1)
class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        
        def count(d):
            ans, cur = 1, 0
            for num in nums:
                cur += num
                if cur > d:
                    ans += 1
                    cur = num
            return ans
        
        l, r = max(nums), sum(nums) + 1
        while l < r:
            mid = l + (r - l) // 2
            if count(mid) <= m:
                r = mid
            else:
                l = mid + 1
        return l